arrow-fx-test/src/main/kotlin/arrow/test/laws/BracketLaws.kt

```diff
@@ -154,15 +168,62 @@ object BracketLaws {
         .equalUnderTheLaw(just(expected), EQ)
     }
 
-  fun <F> Bracket<F, Throwable>.guaranteeMustRunFinalizerOnError(
+  fun <F> Bracket<F, Throwable>.guaranteeCaseMustRunFinalizerOnError(
     EQ: Eq<Kind<F, Int>>
   ): Unit =
-    forAll(Gen.int(), Gen.throwable()) { i, t ->
+    forAll(Gen.int(), Gen.throwable().raiseError<F, Int, Throwable>(this)) { i, fe ->
       val msg = AtomicIntW(0)
-      raiseError<Int>(t)
-        .guarantee(unit().map { msg.value = i })
+      fe
+        .guaranteeCase { unit().map { msg.value = i } }
         .attempt()
         .map { msg.value }
         .equalUnderTheLaw(just(i), EQ)
     }
+
+  fun <F> Bracket<F, Throwable>.onErrorMustRunFinalizerOnError(
+    EQ: Eq<Kind<F, Int>>
+  ): Unit =
+    forAll(
+      Gen.throwable().raiseError<F, Int, Throwable>(this),
+      Gen.int()
+    ) { fe: Kind<F, Int>, i: Int ->
+      val msg = AtomicIntW(0)
+      fe
+        .onError { unit().map { msg.value = i } }
+        .attempt()
+        .map { msg.value }
+        .equalUnderTheLaw(just(i), EQ)
+    }
+
+  fun <F> Bracket<F, Throwable>.bracketShouldBeStackSafe(
+    iterations: Int,
+    EQ: Eq<Kind<F, Int>>
+  ) {
+
+    fun ioBracketLoop(i: Int): Kind<F, Int> =
+      unit().bracket(use = { just(i + 1) }, release = { unit() }).flatMap { ii ->
+        if (ii < iterations) ioBracketLoop(ii)
+        else just(ii)
+      }
+
+    forFew(100, Gen.int().applicativeError(this)) { fa ->
```

Comment:
We run this over `forAll(Gen.create { Unit }) {` in `MonadDeferLaws.kt` and we do not run them in `forAll` in `ConcurrentLaws.kt`.

I think `forAll(Gen.create { Unit })` does a single iteration due to `Unit`.
These tests are quite slow, so running them 100x will take a considerable amount of time per datatype.

Reply:
`Gen.int().applicativeError(this)` is not meaning-full for testing stack safety since all errors will short-circuit the chain and by-pass the stack-safety issue.

We should probably replace all these instances of `applicativeError` with `GenK`
Maybe @1Jajen1 can give some more insights on this?

Reply:
> `Gen.int().applicativeError(this)` is not meaning-full for testing stack safety since all errors will short-circuit the chain and by-pass the stack-safety issue.
> 
> We should probably replace all these instances of `applicativeError` with `GenK`
> Maybe @1Jajen1 can give some more insights on this?

`MonadError` has laws which say `raiseError(e).flatMap(f) == raiseError(e)` so if `Gen.applicativeError` would only generate errors it'd not do anything. But `applicativeError` is defined as `Gen.oneOf` which does both states. It's still a bad gen, but it should generate both states. This test should however fail as soon as an error is generated!

Reply:
> I think forAll(Gen.create { Unit }) does a single iteration due to Unit.

Is that really the case? o.O I mean I know kotlintest uses sequences internally to model the random values, so this is possible, but that's weird nevertheless...

Reply:
I haven't verified that statement @1Jajen1! I simply made that assumption based on the speed of the tests.

Reply:
The assumption makes sens if they implemented `Gen.create { x } == Gen(sequenceOf(x))` or similar. Either way I'd say this is quite weird if one specifically wants to test x-times, so we should definitly test that somehow...